In [1]:
import requests
from bs4 import BeautifulSoup

import numpy as np
import pandas as pd

%matplotlib inline
import matplotlib

In [2]:
url = "http://books.toscrape.com/"

In [3]:
response = requests.get(url, headers={"Accept-Language":"en-US"})
soup = BeautifulSoup(response.content, "html.parser")

< article class="product_pod">

In [6]:
books_html  = soup.find_all("article", class_="product_pod")

## Parsing _one_ book

In [15]:
type(books_html)
len(books_html)

20

In [43]:
#Title
books_html[0].find("h3").find("a").attrs
book_title = books_html[0].find("h3").find("a").attrs["title"]
float(books_html[0].find("p", class_="price_color").string[1:])

51.77

In [53]:
books_html[0].find("p", class_="star-rating")['class'][1]

'Three'

In [54]:
book_stars_html = books_html[0].find("p", class_="star-rating")
book_stars_html

<p class="star-rating Three">
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
</p>

In [44]:
books_html[0].find("p", class_="star-rating")

<p class="star-rating Three">
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
<i class="icon-star"></i>
</p>

In [56]:
def parse_rating(rating_classes):
    
    if rating_classes == 'One':
        return 1
    
    elif rating_classes == 'Two':
        return 2
    
    elif rating_classes == 'Three':
        return 3 
    
    elif rating_classes == 'Four':
        return 4
    
    elif rating_classes == 'Five':
        return 5   

    return 0

In [61]:
parse_rating(books_html[3].find("p", class_="star-rating")['class'][1])

4

Title => ["A light in the attic", "Tipping the Velvet", ...]

Price => [51.77, 53.74, ...]

Rating => [3, 1, ...]

In [74]:
books_dict={'Title':[], 'Price':[], 'Rating':[]}


for book in books_html:
    title = book.find("h3").find("a").attrs["title"]
    price = float(book.find("p", class_="price_color").string[1:])
    rating = parse_rating(book.find("p", class_="star-rating")['class'][1])
    
    books_dict['Title'].append(title) 
    books_dict['Price'].append(price)
    books_dict['Rating'].append(rating)

#book_title = books_html[0].find("h3").find("a").attrs["title"]

In [77]:
books_df = pd.DataFrame.from_dict(books_dict)

In [78]:
from nbresult import ChallengeResult

result = ChallengeResult('books',
    books_dict=books_dict,
    columns=books_df.columns,
    title=str(books_df.loc[0,'Title']),
    price=books_df.loc[0,'Price'],
    rating=books_df.loc[0,'Rating']
)
result.write()
print(result.check())

============================= test session starts ==============================
platform linux -- Python 3.8.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /home/nandosoq/.pyenv/versions/lewagon/bin/python3
cachedir: .pytest_cache
rootdir: /home/nandosoq/code/Nandosoq/data-challenges/02-Data-Toolkit/02-Data-Sourcing/02-Scraping
plugins: anyio-3.2.1, dash-1.21.0
collecting ... collected 6 items

tests/test_books.py::TestBooks::test_books_df_columns_name PASSED        [ 16%]
tests/test_books.py::TestBooks::test_books_df_columns_size PASSED        [ 33%]
tests/test_books.py::TestBooks::test_books_dict_id_defined PASSED        [ 50%]
tests/test_books.py::TestBooks::test_first_book_price PASSED             [ 66%]
tests/test_books.py::TestBooks::test_first_book_rating PASSED            [ 83%]
tests/test_books.py::TestBooks::test_first_book_title PASSED             [100%]

============================== 6 passed in 0.30s ===============================


💯 You can commit your code:

git add te